django/db/models/sql/compiler.py

```diff
@@ -903,17 +905,102 @@ def __init__(self, *args, **kwargs):
         self.return_id = False
         super(SQLInsertCompiler, self).__init__(*args, **kwargs)
 
-    def placeholder(self, field, val):
+    def field_as_sql(self, field, val):
+        """
+        Take a field, and a value intended to be saved on that field, and
+        return placeholder SQL and accompanying params. Checks for raw values,
+        expressions and fields with get_placeholder() defined in that order.
+        """
+
         if field is None:
             # A field value of None means the value is raw.
-            return val
+            sql, params = val, []
+
+        elif hasattr(val, 'as_sql'):
+            # This is an expression, let's compile it.
+            sql, params = self.compile(val)
+
         elif hasattr(field, 'get_placeholder'):
             # Some fields (e.g. geo fields) need special munging before
             # they can be inserted.
-            return field.get_placeholder(val, self, self.connection)
+            sql, params = field.get_placeholder(val, self, self.connection), [val]
+
         else:
             # Return the common case for the placeholder
-            return '%s'
+            sql, params = '%s', [val]
+
+        # The following hook is only used by Oracle Spatial, which sometimes
+        # needs to yield 'NULL' and [] as its placeholder and params instead
+        # of '%s' and [None]. The 'NULL' placeholder is produced earlier by
+        # OracleOperations.get_geom_placeholder(). The following line removes
+        # the corresponding None parameter. See ticket #10888.
+        params = self.connection.ops.modify_insert_params(sql, params)
+
+        return sql, params
+
+    def prepare_value(self, field, value):
+        """
+        Prepare a value to be used in a query, by resolving it if it is an
+        expression, and otherwise calling the field's get_db_prep_save().
+        """
+
+        def find_fs(expr):
```

Comment:
Yeah, it's clear now that this approach would have broken for anything other than simple `CombinedExpression`s. I've added a test to confirm and implemented `BaseExpression.flatten()`.

I also implemented a corresponding method on `F` simply yielding `self`.

I added the explicit check because there was an existing test for this case. Without an explicit check, what was originally a `TypeError` raised in (IIRC) `get_db_prep_save()` turned into a more cryptic error at the database level. See the change in `test_object_update_unsaved_objects()` in `tests/expressions.py`.